Mesh dataset 3D rendering

vcloarec opened the pull request:
Mesh dataset 3D rendering

Samweli committed:
added abstract temporal class

Samweli committed:
updates on tests and abstract temporal class

Samweli committed:
updates on qgsabstracttemporal sip and tests

Samweli committed:
update qgsabstracttemporal test

Samweli committed:
updated docs

Samweli committed:
moved qgsabstracttemporal class to qgstemporalrangeobject class as per review suggestions

Samweli committed:
edits per review on qgs map canvas

Samweli committed:
remove access of private members from qgstemporalrangeobject

Samweli committed:
updated render context test and fix travis tests

Samweli committed:
added new temporal property class

Samweli committed:
added tests for temporal property

Samweli committed:
added new map layer temporal properties sub classes with tests

Samweli committed:
updates on map layers temporal properties classes

Samweli committed:
fix travis tests

Samweli committed:
add read and write xml tests

Samweli committed:
update raster layer temporal properties tests

Commit:
Add a check for nullptr on mGatherer in QgsFeatureFilterModel::updateCompleter.
Also avoid relying on signal to slot call order conservation to make sure that gathererThreadFinished is called after updateCompleter :
  - gathererThreadFinished is disconnected from finished signal emitted at the end of QgsFieldExpressionValuesGatherer::run
  - gathererThreadFinished is instead called explicitly at the end of updateCompleter
  - when QgsFieldExpressionValuesGatherer::run is stopped manually, rely on the connection to QgsFieldExpressionValuesGatherer::deleteLater to clean mGatherer.

This should fix an observed bug where a crash happened at the begining of updateCompleter because mGatherer was null.

mhugent committed:
DXF export: handle line breaks in label text. Line breaks in TEXT or MTEXT makes dxf viewers hang or crash

mhugent committed:
astyle

Commit:
Fix style.

Samweli committed:
removed unwanted temporal range object method

nyalldawson committed:
Add API to QgsFieldModel/QgsFieldComboBox to manually set the fields
to show in the widget (i.e. when no layer is available)

nyalldawson committed:
[processing] Fix incorrect error about "cannot load dependant fields"
when a field parameter is linked to a feature source parameter which
has the "selected features only" checked

nyalldawson committed:
[FEATURE][processing] New parameter type for datetime (or date, or time) values

Adds a new parameter type QgsProcessingParameterDateTime for handling
date/datetime/time values

nyalldawson committed:
Add QgsDateTimeEdit subclasses for specific date and time widgets

Adds QgsDateEdit and QgsTimeEdit subclasses. (All the heavy lifting
with these is done by Qt itself, hence no extra unit tests!)

nyalldawson committed:
Allow null string to be customised for QgsDateTimeEdit

E.g. for processing use this needs to be "Not specified"

elpaso committed:
Move jsonformatter to own dir, add license and src

Fixes #34586

nyalldawson committed:
Add support for reading/writing QDateTime, QDate and QTime via QgsXmlUtils

nyalldawson committed:
Fix duplicate signals emitted by QgsDateTimeEdit when changed

nyalldawson committed:
Add widget wrapper for datetime parameters

nyalldawson committed:
Mark some internal methods as private

nyalldawson committed:
Fix crash in test

vcloarec committed:
dataset scalar 3D

nyalldawson committed:
Merge pull request #34040 from Samweli/temporal_api

Temporal API

nyalldawson committed:
Minor temporal properties cleanups and dox improvements

nyalldawson committed:
Purge totally completely horribly outdated QgsRasterLayer docs

mhugent committed:
Merge pull request #34574 from mhugent/dxf_label_linebreaks

DXF export: handle line breaks in label text

PeterPetrik commented on src/3d/mesh/qgsmesh3dentity_p.cpp:
```diff
@@ -27,14 +27,16 @@
+  mSymbol( symbol )
+{}
+
+QgsMeshDataset3dEntity::QgsMeshDataset3dEntity( const Qgs3DMapSettings &map,
```
have you run ./prepare_commit.sh , the formatting looks strange

PeterPetrik commented on src/3d/mesh/qgsmesh3dentity_p.cpp:
```diff
@@ -43,59 +45,59 @@ void QgsMesh3dEntity::build()
-  {
-    fakeScalarMag[i] = mTriangularMesh.vertices().at( i ).z();
-  }
+  QgsMeshLayer *l = layer();
```
one char variables are FORTRAN, best to use some longer  variable names , or use layer()

PeterPetrik commented on src/3d/mesh/qgsmesh3dentity_p.cpp:
```diff
@@ -43,59 +45,59 @@ void QgsMesh3dEntity::build()
-  addComponent( mMaterial );
+  if ( mSymbol.renderingStyle() == QgsMesh3DSymbol::ColorRamp2DRendering && layer() )
+  {
+    QgsMeshRendererSettings rendererSettings = layer()->rendererSettings();
```
const

PeterPetrik commented on src/3d/mesh/qgsmesh3dentity_p.cpp:
```diff
@@ -43,59 +45,59 @@ void QgsMesh3dEntity::build()
+  if ( mSymbol.renderingStyle() == QgsMesh3DSymbol::ColorRamp2DRendering && layer() )
+  {
+    QgsMeshRendererSettings rendererSettings = layer()->rendererSettings();
+    QgsMeshDatasetIndex datasetIndex = rendererSettings.activeScalarDataset();
```
const

PeterPetrik commented on src/3d/mesh/qgsmesh3dentity_p.cpp:
```diff
@@ -43,59 +45,59 @@ void QgsMesh3dEntity::build()
 QgsMesh3dTerrainTileEntity::QgsMesh3dTerrainTileEntity( const Qgs3DMapSettings &map,
-    const QgsTriangularMesh triangularMesh,
-    const QgsRectangle &extent,
+    QgsMeshLayer *meshLayer,
```
const ? do you need to modify it?

vcloarec replied:
This class has been deeply modified by refactoring. Now, it is derived from `QgsMesh3dEntity` and `QgsTerrainTileEntity`.
And the similar class `QgsMeshDataset3dEntity` is derived from `QgsMesh3dEntity` and `Qt3DCore::QEntity`

PeterPetrik replied:
I mean if it shouldn't be `const QgsMeshLayer *meshLayer`

vcloarec replied:
this parameter is used by the constructor of `QgsMapLayerRef` that don't support `const QgsMeshLayer*` but supports `QgsMeshLayer*`

PeterPetrik replied:
ok

PeterPetrik commented on src/3d/mesh/qgsmesh3dentity_p.cpp:
```diff
@@ -43,59 +45,59 @@ void QgsMesh3dEntity::build()
 void QgsMesh3dTerrainTileEntity::buildGeometry()
 {
   Qt3DRender::QGeometryRenderer *mesh = new Qt3DRender::QGeometryRenderer;
+  QgsMeshLayer *l = layer();
```
1 char variables...

PeterPetrik commented on src/3d/mesh/qgsmesh3dentity_p.h:
```diff
@@ -43,60 +43,61 @@ class QgsMesh3DSymbol;
+    virtual void applyMaterial() = 0;
+};
+
+//! Entity that handles rendering of dataset mesh
```
remove mesh. in comment

PeterPetrik commented on src/3d/mesh/qgsmesh3dentity_p.h:
```diff
@@ -43,60 +43,61 @@ class QgsMesh3DSymbol;
     QgsMesh3dTerrainTileEntity( const Qgs3DMapSettings &map,
-                                const QgsTriangularMesh triangularMesh,
-                                const QgsRectangle &extent,
+                                QgsMeshLayer *meshLayer,
```
const?

vcloarec replied:
`const` seems not compatible with the `QgsMapLayerRef` constructor that uses the parameter

PeterPetrik replied:
ok

PeterPetrik commented on src/3d/mesh/qgsmesh3dgeometry_p.cpp:
```diff
@@ -56,11 +54,8 @@ static QByteArray createPlaneVertexData( const QgsTriangularMesh &mesh,
-
-    QVector3D normal = normales.at( i ).normalized();
-    normal = QVector3D( normal.x(), -normal.y(), normal.z() * vertScale );
+    QVector3D normal = normales.at( i );
```
normals (typo)

PeterPetrik commented on src/3d/mesh/qgsmesh3dgeometry_p.cpp:
```diff
@@ -71,27 +66,58 @@ static QByteArray createPlaneVertexData( const QgsTriangularMesh &mesh,
 
-  QVector<QVector3D> normales = mesh.vertexNormals( vertScale );
+  //Calculate normales with Z value equal to verticaleMagnitude
+  QVector<QVector3D> normales( nVerts );
```
normals

PeterPetrik commented on src/3d/mesh/qgsmesh3dgeometry_p.cpp:
```diff
@@ -71,27 +66,58 @@ static QByteArray createPlaneVertexData( const QgsTriangularMesh &mesh,
-  QVector<QVector3D> normales = mesh.vertexNormals( vertScale );
+  //Calculate normales with Z value equal to verticaleMagnitude
+  QVector<QVector3D> normales( nVerts );
+  for ( const auto &face : mesh.triangles() )
```
don't we have such function in QgsMeshUtils namespace? eventually I would extract this to new function for better maintanance/testing/usability

vcloarec replied:
Yes you are right.
My motto would be : why make it simple when I can make it complicated :confused:
You can see I had used two different way to make the same thing...
So I have added a function to QgsMeshLayerUtils and use it, and everything seems to be allright.

PeterPetrik commented on src/3d/mesh/qgsmesh3dgeometry_p.cpp:
```diff
@@ -71,27 +66,58 @@ static QByteArray createPlaneVertexData( const QgsTriangularMesh &mesh,
+  }
 
   // Populate a buffer with the interleaved per-vertex data with
   // vec3 pos, vec2 texCoord, vec3 normal, vec4 tangent
```
update comment?

PeterPetrik commented on src/3d/mesh/qgsmesh3dgeometry_p.cpp:
```diff
@@ -141,199 +166,472 @@ static QByteArray createPlaneIndexData( const QgsTriangularMesh &mesh )
+{
+  const int faces = mesh.triangles().count();
+  const quint32 indices = static_cast<quint32>( 3 * activeFaceCount );
+  Q_ASSERT( indices < std::numeric_limits<quint32>::max() );
```
hmm... probably not needed

PeterPetrik commented on src/3d/mesh/qgsmesh3dgeometry_p.cpp:
```diff
@@ -141,199 +166,472 @@ static QByteArray createPlaneIndexData( const QgsTriangularMesh &mesh )
+
+static QByteArray createDatasetIndexData( const QgsTriangularMesh &mesh, const QgsMeshDataBlock &mActiveFaceFlagValues, int activeFaceCount )
+{
+  const int faces = mesh.triangles().count();
```
triangleCount?

PeterPetrik commented on src/3d/mesh/qgsmesh3dgeometry_p.cpp:
```diff
@@ -141,199 +166,472 @@ static QByteArray createPlaneIndexData( const QgsTriangularMesh &mesh )
-        float vertScale )
+        const QgsVector3D &origin,
+        float vertScale,
+        bool verticalRelative = false )
```
can you add docs for the verticalRelative attribute?

PeterPetrik commented on src/3d/mesh/qgsmesh3dgeometry_p.cpp:
```diff
@@ -141,199 +166,472 @@ static QByteArray createPlaneIndexData( const QgsTriangularMesh &mesh )
+{}
+
+QgsMeshDataset3dGeometry::QgsMeshDataset3dGeometry(
+  QgsMeshLayer *layer,
```
const?

PeterPetrik commented on src/3d/mesh/qgsmesh3dgeometry_p.cpp:
```diff
@@ -141,199 +166,472 @@ static QByteArray createPlaneIndexData( const QgsTriangularMesh &mesh )
+  int dataOnVerticesCount = triangularMesh.vertices().count();
+
+  //***** extract the scalar dataset used to render vertical magintude of geometry
+  //define the vertical magitude datasetIndex
```
magnitude

PeterPetrik commented on src/3d/mesh/qgsmesh3dgeometry_p.cpp:
```diff
@@ -141,199 +166,472 @@ static QByteArray createPlaneIndexData( const QgsTriangularMesh &mesh )
+  QgsMeshDatasetIndex verticalMagDatasetIndex( mVerticalGroupDatasetIndex, verticalDataSetIndexNumber );
+
+  //define the active face for vertical magnitude, the inactive faces will not be rendered
+  activeFaceFlagValues = layer->dataProvider()->areFacesActive( verticalMagDatasetIndex, 0, nativeMesh.faces.count() );
```
dataProvider() doesn't need to support active faces, in that case all faces are active

vcloarec replied:
I don't understand your comment... IN all cases, that do the job, only actives faces are rendered

PeterPetrik replied:
ok

PeterPetrik commented on src/3d/mesh/qgsmesh3dgeometry_p.cpp:
```diff
@@ -141,199 +166,472 @@ static QByteArray createPlaneIndexData( const QgsTriangularMesh &mesh )
+                          QgsMeshRendererScalarSettings::NeighbourAverage
+                        );
+    }
+  }
```
isn't the code same as the one above for vertical magnitude? In that case maybe worth to create helper function in QgsMeshLayerUtils or somewhere here to extract & interpolate date to vertices

vcloarec replied:
Yes you are right.
My motto would be : why make it simple when I can make it complicated :confused:
You can see I had used two different way to make the same thing...
So I have added a function to QgsMeshLayerUtils and use it, and everything seems to be allright.

PeterPetrik commented on src/3d/mesh/qgsmesh3dgeometry_p.cpp:
```diff
@@ -141,199 +166,472 @@ static QByteArray createPlaneIndexData( const QgsTriangularMesh &mesh )
+  QgsTriangularMesh triangularMesh = *layer->triangularMesh();
+
+  const int stride = ( 3 /*position*/ +
                        3 /*normale*/ ) * sizeof( float );
```
normal

PeterPetrik commented on src/3d/mesh/qgsmesh3dgeometry_p.cpp:
```diff
@@ -141,199 +166,472 @@ static QByteArray createPlaneIndexData( const QgsTriangularMesh &mesh )
+  indexAttribute->setVertexBaseType( QAttribute::UnsignedInt );
+  indexAttribute->setBuffer( buffer );
+
+  // Each primitive has 3 vertives
```
vertices

PeterPetrik commented on src/3d/mesh/qgsmesh3dgeometry_p.h:
```diff
@@ -43,34 +46,75 @@ namespace Qt3DRender
 {
-  public:
+  protected:
+    //! Constructor protected because must anly be called from derived classes
```
1) typo in comment, maybe you do not need comment at all

PeterPetrik commented on src/3d/mesh/qgsmesh3dgeometry_p.h:
```diff
@@ -43,34 +46,75 @@ namespace Qt3DRender
+    void prepareVerticesNormalAttribute( Qt3DRender::QBuffer *buffer, int count, int stride, int offset );
+    void prepareIndexesAttribute( Qt3DRender::QBuffer *buffer, int count );
+
+    QgsMapLayerRef mLayerRef;
```
maybe private: section?

vcloarec replied:
members used by derived class.

PeterPetrik replied:
ok

PeterPetrik commented on src/3d/mesh/qgsmesh3dgeometry_p.h:
```diff
@@ -43,34 +46,75 @@ namespace Qt3DRender
+/**
+ * \ingroup 3d
+ * Creates attributes and vertex/index buffers for a mesh layer that renders the dataset
+ * \since QGIS 3.14
```
it is in _p.h file, so I assume it is private/implementation detail. do not need to add docs or \since or \ingroup..

PeterPetrik commented on src/3d/mesh/qgsmesh3dgeometry_p.h:
```diff
@@ -43,34 +46,75 @@ namespace Qt3DRender
-    QgsVector3D mOrigin;
-    QgsRectangle mExtent;
-    float mVertScale;
+    // Returns the number of active faces
```
missing ! after //

PeterPetrik commented on src/3d/mesh/qgsmesh3dgeometry_p.h:
```diff
@@ -43,34 +46,75 @@ namespace Qt3DRender
+    int extractDataset( QVector<double> &verticaleMagnitude, QVector<double> &scalarMagnitude, QgsMeshDataBlock &verticalActiveFaceFlagValues );
+    void prepareVerticesDatasetAttribute( Qt3DRender::QBuffer *buffer, int count, int stride, int offset );
+
+    bool mIsVerticalMagnitudeRelative;
```
private: section?

PeterPetrik commented on src/3d/mesh/qgsmesh3dgeometry_p.h:
```diff
@@ -43,34 +46,75 @@ namespace Qt3DRender
+/**
+ * \ingroup 3d
+ * Creates attributes and vertex/index buffers for a mesh layer that renders terrain
+ * \since QGIS 3.14
```
no need for \since and \ingroup

PeterPetrik commented on src/3d/mesh/qgsmesh3dmaterial_p.cpp:
```diff
@@ -77,10 +77,11 @@ class ColorRampTextureGenerator: public Qt3DRender::QTextureImageDataGenerator
 
       dataPtr->setData( data, sizeof( float ) ); //size is the size of the type, here float
 
+      qDebug() << "Color ramp texture created";
```
better to use qgs alternatives (QgsDebugMessageLvl or so) if needed at all

vcloarec replied:
I've forgotten to remove that ...

PeterPetrik commented on src/3d/mesh/qgsmesh3dmaterial_p.cpp:
```diff
@@ -100,8 +101,13 @@ class ColorRampTextureGenerator: public Qt3DRender::QTextureImageDataGenerator
       {
-        if ( colorItemList.at( i ).color != otherColorItemList.at( i ).color ||
-             colorItemList.at( i ).value != otherColorItemList.at( i ).value )
+        QColor color = colorItemList.at( i ).color;
```
const color ?

PeterPetrik commented on src/3d/mesh/qgsmesh3dmaterial_p.cpp:
```diff
@@ -100,8 +101,13 @@ class ColorRampTextureGenerator: public Qt3DRender::QTextureImageDataGenerator
-        if ( colorItemList.at( i ).color != otherColorItemList.at( i ).color ||
-             colorItemList.at( i ).value != otherColorItemList.at( i ).value )
+        QColor color = colorItemList.at( i ).color;
+        QColor otherColor = otherColorItemList.at( i ).color;
```
const ?

PeterPetrik commented on src/3d/mesh/qgsmesh3dmaterial_p.cpp:
```diff
@@ -148,12 +158,30 @@ QgsMesh3dMaterial::QgsMesh3dMaterial( const QgsMesh3DSymbol &symbol ):
+    switch ( mMagnitudeType )
+    {
+      case QgsMesh3dMaterial::ZValue:
+        // if the color shading is done with the Z value of vertices, the color ramhave top adapted with verticale scale
```
, the colorramp has

PeterPetrik commented on src/3d/mesh/qgsmesh3dmaterial_p.cpp:
```diff
@@ -148,12 +158,30 @@ QgsMesh3dMaterial::QgsMesh3dMaterial( const QgsMesh3DSymbol &symbol ):
+        colorRampTexture->addTextureImage( new ColorRampTexture( mSymbol.colorRampShader(), mSymbol.verticalScale() ) );
+        break;
+      case QgsMesh3dMaterial::ScalarDataSet:
+        // if the color shading is done with scalar dataset, no verticale scale to use
```
, no vertical scale

PeterPetrik commented on src/3d/mesh/qgsmesh3dmaterial_p.cpp:
```diff
@@ -148,12 +158,30 @@ QgsMesh3dMaterial::QgsMesh3dMaterial( const QgsMesh3DSymbol &symbol ):
+        // if the color shading is done with scalar dataset, no verticale scale to use
+        colorRampTexture->addTextureImage( new ColorRampTexture( mSymbol.colorRampShader(), 1 ) );
+        break;
+    }
```
and the case with relative scalar data set?

vcloarec replied:
relative is for the vertical magnitude used for geometry, no for the scalar magnitude used for color shading.
Here, it the scalar color shading rendering that is handle.

PeterPetrik replied:
ok

PeterPetrik commented on src/3d/mesh/qgsmesh3dmaterial_p.cpp:
```diff
@@ -168,7 +196,7 @@ void QgsMesh3dMaterial::configure()
 
   //Load shader programs
-  QUrl urlVert( QStringLiteral( "qrc:/shaders/mesh/mesh.vert" ) );
+  QUrl urlVert = QUrl( QStringLiteral( "qrc:/shaders/mesh/mesh.vert" ) );
```
what was wrong on the original line?

vcloarec replied:
nothing, its has undergone a more deeper modification and a come back (bad) to initial...

PeterPetrik commented on src/3d/mesh/qgsmesh3dmaterial_p.cpp:
```diff
@@ -178,19 +206,20 @@ void QgsMesh3dMaterial::configure()
   QColor meshColor = mSymbol.singleMeshColor();
   mTechnique->addParameter( new Qt3DRender::QParameter( "meshColor", QVector4D( meshColor.redF(), meshColor.greenF(), meshColor.blueF(), 1.0f ) ) );
-  mTechnique->addParameter( new Qt3DRender::QParameter( "verticaleScale", float( mSymbol.verticaleScale() ) ) );
+  mTechnique->addParameter( new Qt3DRender::QParameter( "isScalarMagnitude", ( mMagnitudeType == QgsMesh3dMaterial::ScalarDataSet ) ) );
```
I would put there enum if possible, what if you will add another magnitude type?

vcloarec replied:
For now, there is only one another value for this enum. If another magnitude is added in the future, the behavior will be the same : if the value is not `ScalarDataSet`, the shader will not use the scaler magnitude, but the Z value of the vertices.
If another behavior is needed in the shader (if... I am  not sure), there will be the need to modify this line and the shader.

PeterPetrik replied:
i got it, but if that is likely to happen, I would pass enum to shared right now, so extension in the future is super easy

vcloarec replied:
After read again, I think I have misunderstood what you mean here, and a am not sure to understand. Do you want to pass the enum in the parameter ?
If it is that, the parameter is for the shader, and  the shader language doesn't support the enum. We can pass `int`, but it is less explicit, that why I prefer a bool that is really explicit to what we want. Furthermore, if we use `int`, we will not be sure, in the future, that the meaning of the `int` int the shader will be still consistent with the meaning of the parameter.

PeterPetrik commented on src/3d/mesh/qgsmeshterraingenerator.cpp:
```diff
@@ -29,18 +29,21 @@
+  if ( !layer )
+    return nullptr;
+
+  auto entity = new QgsMesh3dTerrainTileEntity( terrain()->map3D(), layer, mSymbol, mNode->tileId(), parent );
```
do not use auto

PeterPetrik commented on src/3d/qgs3dmapscene.cpp:
```diff
@@ -592,6 +592,12 @@ void Qgs3DMapScene::addLayerEntity( QgsMapLayer *layer )
+
+  if ( layer->type() == QgsMapLayerType::MeshLayer )
+  {
+    connect( layer, &QgsMapLayer::rendererChanged, this, &Qgs3DMapScene::onLayerRenderer3DChanged );
```
this signal/slot looks quite generic for me and not connected to mesh layer... why other layers do not need it ?

vcloarec replied:
Before this PR, only vector layer need this. Now there is too mesh layer. I don't now the reason why it is needed for vector layer, but it is needed for mesh layer when the dataset or the color ramp shader of the contour change. 
It is a quite heavy operation because, the entity need to be rebuilt and with this connection, all the times the 2D renderer changes.
I am afraid that for other layer type, it could be heavy task for nothing.

PeterPetrik replied:
ok

PeterPetrik commented on src/3d/qgs3dmapscene.cpp:
```diff
@@ -613,6 +619,11 @@ void Qgs3DMapScene::removeLayerEntity( QgsMapLayer *layer )
+
+  if ( layer->type() == QgsMapLayerType::MeshLayer )
+  {
+    disconnect( layer, &QgsMapLayer::rendererChanged, this, &Qgs3DMapScene::onLayerRenderer3DChanged );
```
this signal/slot looks quite generic for me and not connected to mesh layer... why other layers do not need it ?

vcloarec replied:
same

PeterPetrik replied:
ok

PeterPetrik commented on src/3d/symbols/qgsmesh3dsymbol.cpp:
```diff
@@ -168,3 +172,23 @@ void QgsMesh3DSymbol::setRenderingStyle( const QgsMesh3DSymbol::RenderingStyle &
   mRenderingStyle = coloringType;
 }
+
+int QgsMesh3DSymbol::verticalDatasetGroupIndex() const
```
I think somewhere (not sure where best) is to add proper documentation that what this vertical dataset is and how it differs from other dataset used for colouring

vcloarec replied:
The doc in the *.h file is not enough ?

PeterPetrik replied:
ok

PeterPetrik commented on src/3d/symbols/qgsmesh3dsymbol.h:
```diff
@@ -48,7 +49,22 @@ class _3D_EXPORT QgsMesh3DSymbol : public QgsAbstract3DSymbol
+    {
+      //! Use the Z value of the vertices
+      VerticesZValue = 0,
+      //! Render the mesh with a color ramp
```
color ramp?

vcloarec replied:
no, I am wrong (bad copy/paste...)

PeterPetrik commented on src/3d/symbols/qgsmesh3dsymbol.h:
```diff
@@ -48,7 +49,22 @@ class _3D_EXPORT QgsMesh3DSymbol : public QgsAbstract3DSymbol
+    };
+
+    /**
+     * Which value to render the Z value of the mesh
```
... how to calculate Z value of vertices for 3D scene

PeterPetrik commented on src/3d/symbols/qgsmesh3dsymbol.h:
```diff
@@ -48,7 +49,22 @@ class _3D_EXPORT QgsMesh3DSymbol : public QgsAbstract3DSymbol
+     *
+     * \since QGIS 3.14
+     */
+    enum ZvalueType
```
ZValueType

PeterPetrik commented on src/3d/symbols/qgsmesh3dsymbol.h:
```diff
@@ -197,20 +213,55 @@ class _3D_EXPORT QgsMesh3DSymbol : public QgsAbstract3DSymbol
     void setRenderingStyle( const QgsMesh3DSymbol::RenderingStyle &textureType );
 
+    /**
+     * Returns the index og the dataset group that will be used to render the vertical of the 3D mesh
```
of

PeterPetrik commented on src/3d/symbols/qgsmesh3dsymbol.h:
```diff
@@ -197,20 +213,55 @@ class _3D_EXPORT QgsMesh3DSymbol : public QgsAbstract3DSymbol
+    int verticalDatasetGroupIndex() const;
+
+    /**
+     * Sets the index og the dataset group that will be used to render the vertical of the 3D mesh
```
.. of ..

PeterPetrik commented on src/3d/symbols/qgsmesh3dsymbol.h:
```diff
@@ -197,20 +213,55 @@ class _3D_EXPORT QgsMesh3DSymbol : public QgsAbstract3DSymbol
+    void setVerticalDatasetGroupIndex( int verticalDatasetGroupIndex );
+
+    /**
+     * Returns if the verticale magnitude is relative to the mesh vertices Z value
```
vertical ...

PeterPetrik commented on src/3d/symbols/qgsmesh3dsymbol.h:
```diff
@@ -197,20 +213,55 @@ class _3D_EXPORT QgsMesh3DSymbol : public QgsAbstract3DSymbol
+    bool isVerticalMagnitudeRelative() const;
+
+    /**
+     * Sets if the verticale magnitude is relative to the mesh vertices Z value
```
vertical ...

PeterPetrik commented on src/3d/symbols/qgsmesh3dsymbol.h:
```diff
@@ -197,20 +213,55 @@ class _3D_EXPORT QgsMesh3DSymbol : public QgsAbstract3DSymbol
+
   private:
 
+    //old settings
```
old? not sure if I understand, maybe remove the comment

vcloarec replied:
Well, I have put that comment, to remember what to do with the "old" implementation of the mesh 3D rendering. I don't think there are needed anymore, so I remove its.

vcloarec replied:
Well, I have keep that lines and put "old" to remember what to do with the old implementation of the mesh 3D rendering. Those settings are related to the old implementation that used `QgsMesh3DSymbolEntity` which are still here and used those settings.

PeterPetrik replied:
there is nothing like "old". you cannot stop QGIS to become better. best to remove this comment
```

PeterPetrik commented on src/app/3d/qgsmesh3dsymbolwidget.h:
```diff
@@ -37,10 +37,14 @@ class QgsMesh3dSymbolWidget : public QWidget, private Ui::QgsMesh3dPropsWidget
-    void enableVerticalSetting( bool isEnable );
+    void configureForTerrain();
+    void configureForDataset();
+
```
extra 2 lines

PeterPetrik commented on src/app/mesh/qgsmeshdatasetgrouptreeview.h:
```diff
@@ -200,4 +200,20 @@ class APP_EXPORT QgsMeshDatasetGroupTreeView : public QTreeView
     QgsMeshLayer *mMeshLayer = nullptr; // not owned
 };
 
+class QgsMeshDatasetGroupListModel: public QAbstractListModel
```
is this class somewhere used?

vcloarec replied:
yes, in the mesh 3D symbol widget, to choose the dataset for rendering the Z component.

PeterPetrik replied:
ok

m-kuhn committed:
Merge pull request #34581 from obrix/fix_qgsfeaturefiltermodel_updateCompleter_sync

Small rework on QgsFeatureFilterModel

elpaso committed:
Merge pull request #34612 from elpaso/bugfix-gh34586-jsonformatter

Move jsonformatter to own dir, add license and src

nyalldawson committed:
Accept "proj:..." prefix in QgsCoordinateReferenceSystem::createFromString
instead of just "proj4:..."

nyalldawson committed:
Show formatted WKT definitions in custom projection dialog

Because an unformatted WKT string is just a huge barely decipherable blob...

nyalldawson committed:
Always default to WKT definitions for newly added custom projections

nyalldawson committed:
Show a more helpful message when a custom projection based on WKT
is equivalent to a known projection

vcloarec committed:
after Peter review
.

nyalldawson committed:
[processing] Fix cannot edit/delete/export project based models

Fixes #34627

nyalldawson committed:
Add missing cast for QgsProcessingParameterDateTime

nirvn committed:
[decorations] Fix scale bar label horizontal alignment regression

vcloarec committed:
[FEATURE] mesh simplification (#34532)

[FEATURE] fix #31575 mesh layer speed rendering improvements (LOD)

As mesh layer could have millions of triangles, rendering could be very slow, especially when all the triangles are displayed in the view whereas triangles are too small to be viewed.

For those situations QGIS has an option to simplify the mesh. Simplification leads to one or more simplified mesh that represents levels of detail (LOD). When rendering the mesh, the appropriate level of detail is chosen to have an adequate rendering depending on the view.

elpaso committed:
Fix show/hide link/unlink in relation reference w

widget.

Fixes #34651

3nids committed:
fix browsing counter not reset in feature list (attribute table) (#34501)

the counter was not set to 0 when switching to selected or visible only and if there were no features listed

elpaso committed:
Merge pull request #34652 from elpaso/bugfix-gh34651-relation-reference-embedded-show-hide-link-unlink

Fix show/hide link/unlink in relation reference w

3nids committed:
use Q_ENUM macro in QgsFileWidget (#34665)

this will allow configuring properly the custom widget

vcloarec committed:
fix triangular mesh update

vcloarec committed:
add normals to QgsMeshLayerUtils

vcloarec committed:
add tests

PeterPetrik commented:
@vcloarec i am not sure if it is good idea to merge stuff in master in your fork. Your fork master should be 1-1 with upstream master, otherwise you will end up with quite a trouble when doing PR's to the official upstream master. I propose you to revert the merge and update the master with the upstream master (even by --force)
